src/middleware.rs

```diff
@@ -1,50 +1,93 @@
+use std::iter::successors;
 use std::str::FromStr;
+use std::time::Duration;
 
 use async_recursion::async_recursion;
 use async_trait::async_trait;
-
+use ethers_core::abi::Detokenize;
+use ethers_core::types::{Address, BlockNumber, Selector, TransactionRequest, H160, U256};
 use ethers_core::{
-    abi::{self, Detokenize, ParamType, Token},
-    types::{
-        transaction::eip2718::TypedTransaction, Address, BlockId, BlockNumber, Bytes,
-        NameOrAddress, Selector, TransactionRequest, H160, U256,
-    },
-    utils::{self, hex},
+    abi::{self, ParamType, Token},
+    types::{transaction::eip2718::TypedTransaction, BlockId, Bytes, NameOrAddress},
+    utils::hex,
 };
 use ethers_providers::{ens, erc, Middleware, MiddlewareError};
 use futures_util::try_join;
 use hex::FromHex;
-use reqwest::{Response, Url};
+use reqwest::Url;
 use serde_json::Value;
 
-use crate::{
-    utils::{decode_bytes, dns_encode},
-    CCIPReadMiddlewareError,
-};
+use crate::ccip::handle_ccip;
+use crate::utils::{decode_bytes, dns_encode};
+use crate::CCIPReadMiddlewareError;
 
 #[derive(Debug, Clone)]
 pub struct CCIPReadMiddleware<M> {
-    inner: M,
+    provider: M,
+    ens: Address,
+    reqwest_client: reqwest::Client,
+    max_redirect_attempt: u8,
+}
+
+pub struct CCIPReadMiddlewareBuilder<M> {
+    provider: Option<M>,
     ens: Option<Address>,
+    timeout: Option<Duration>,
+    max_redirect_attempt: Option<u8>,
 }
 
-static MAX_CCIP_REDIRECT_ATTEMPT: u8 = 10;
+impl<M> Default for CCIPReadMiddlewareBuilder<M> {
+    fn default() -> Self {
+        CCIPReadMiddlewareBuilder {
+            provider: None,
+            ens: None,
+            timeout: None,
+            max_redirect_attempt: None,
+        }
+    }
+}
 
-impl<M> CCIPReadMiddleware<M>
-where
-    M: Middleware,
-{
+impl<M: Middleware> CCIPReadMiddlewareBuilder<M> {
+    pub fn with_provider(mut self, provider: M) -> Self {
+        self.provider = Some(provider);
+        self
+    }
+
+    pub fn with_timeout(mut self, timeout: Duration) -> Self {
+        self.timeout = Some(timeout);
+        self
+    }
+
+    pub fn with_max_redirect_attempt(mut self, max_redirect_attempt: u8) -> Self {
+        self.max_redirect_attempt = Some(max_redirect_attempt);
+        self
+    }
+
+    pub fn build(self) -> Result<CCIPReadMiddleware<M>, String> {
+        Ok(CCIPReadMiddleware {
+            provider: self.provider.ok_or("provider is required".to_string())?,
+            ens: self.ens.unwrap_or(ens::ENS_ADDRESS),
+            reqwest_client: reqwest::Client::builder()
+                // TODO: discuss an appropriate default timeout, 4 seconds is a very random decision
```

Comment:
it may be better keep the timeout longer by default. with a quick search found out most of the http request implementations goes far beyond a min by default, but we do not need such big numbers. how about 10 sec?

Reply:
Yeah, 10 seconds sounds good, I'll make the change.